gcnn/train/test_train_validate.py

```diff
@@ -0,0 +1,262 @@
+        )
+        val_mae, val_taskserr, val_taskserr_nodes = validate(val_loader, model)
+        test_rmse = test(test_loader, model)
+        scheduler.step(val_mae)
+        if writer is not None:
+            writer.add_scalar("train error", train_mae, epoch)
+            writer.add_scalar("validate error", val_mae, epoch)
+            writer.add_scalar("test error", test_rmse[0], epoch)
+            for ivar in range(model.num_heads):
+                writer.add_scalar(
+                    "train error of task" + str(ivar), train_taskserr[ivar], epoch
+                )
+        print(
+            f"Epoch: {epoch:02d}, Train MAE: {train_mae:.8f}, Val MAE: {val_mae:.8f}, "
+            f"Test RMSE: {test_rmse[0]:.8f}"
+        )
+        print("Tasks MAE:", train_taskserr)
+
+        total_loss_train.append(train_mae)
+        total_loss_val.append(val_mae)
+        total_loss_test.append(test_rmse[0])
+        task_loss_train_sum.append(train_taskserr)
+        task_loss_val_sum.append(val_taskserr)
+        task_loss_test_sum.append(test_rmse[1])
+
+        task_loss_train.append(train_taskserr_nodes)
+        task_loss_val.append(val_taskserr_nodes)
+        task_loss_test.append(test_rmse[2])
+
+        ###tracking the solution evolving with training
+        if plot_hist_solution:
+            true_values = test_rmse[3]
+            predicted_values = test_rmse[4]
+            visualizer.create_scatter_plots(
+                true_values,
+                predicted_values,
+                output_names=config["Variables_of_interest"]["output_names"],
+                iepoch=epoch,
+            )
+
+    # At the end of training phase, do the one test run for visualizer to get latest predictions
+    test_rmse, test_taskserr, test_taskserr_nodes, true_values, predicted_values = test(
+        test_loader, model
+    )
+
+    ##output predictions with unit/not normalized
+    if config["Variables_of_interest"]["denormalize_output"] == "True":
+        true_values, predicted_values = output_denormalize(
+            config["Variables_of_interest"]["y_minmax"], true_values, predicted_values
+        )
+
+    ######result visualization######
+    visualizer.create_plot_global(
+        true_values,
+        predicted_values,
+        output_names=config["Variables_of_interest"]["output_names"],
+    )
+    visualizer.create_scatter_plots(
+        true_values,
+        predicted_values,
+        output_names=config["Variables_of_interest"]["output_names"],
+    )
+    ######plot loss history#####
+    visualizer.plot_history(
+        total_loss_train,
+        total_loss_val,
+        total_loss_test,
+        task_loss_train_sum,
+        task_loss_val_sum,
+        task_loss_test_sum,
+        task_loss_train,
+        task_loss_val,
+        task_loss_test,
+        model.loss_weights,
+        config["Variables_of_interest"]["output_names"],
+    )
+
+
+def output_denormalize(y_minmax, true_values, predicted_values):
```

Comment:
remove output_denormalize and call it from postprocess.py since it has been moved there.